torch/csrc/autograd/custom_function.h

```diff
@@ -11,4 +11,197 @@ TORCH_API variable_list _wrap_outputs(
   const std::unordered_set<at::TensorImpl*> &dirty_inputs,
   const at::ArrayRef<Variable> raw_outputs,
   const std::shared_ptr<Node> &cdata);
+
+// To use custom autograd operations implement a CFunction subclass with
+// static backward and forward functions
+//
+// forward() can take as many arguments as you want and should return a
+// variable list. Use of any direct Variable arguments will be registered in
+// the graph but no vectors/sets or any other data structures will be traversed.
+//
+// backward() will be given a variable list containing as many Variables as
+// there were outputs from forward. It should return as many Variables as there
+// were inputs with each of them containing the gradient w.r.t. its
+// corresponding input
+//
+// For example:
+// class MyFunction : public CFunction<MyFunction> {
+//   public:
+//   static variable_list forward(AutogradContext *ctx, int n, Variable var);
+//
+//   static variable_list backward(AutogradContext *ctx, variable_list grad_output);
+// };
+// To use MyFunction
+// Variable x;
+// MyFunction::apply(6, x);
+template <class T>
+struct TORCH_API Function {
+  template<typename... Args>
+  static variable_list apply(Args&&... args);
+};
+
+struct AutogradContext {
+  std::vector<torch::autograd::SavedVariable> saved_variables;
+
+  std::unordered_set<at::TensorImpl*> non_differentiable;
+  std::unordered_set<at::TensorImpl*> dirty_inputs;
```

Comment:
Fields in the struct should have a trailing underscore. This is especially important because these count as the public API (though, in your next patch, you make them private, so it is less of a big deal.)